### ISSUE TITLE
RNN for multi class classification

### ISSUE THREAD

**Issue by danigamba** [facebookarchive/caffe2#1679]
Hello,
I'm trying to build and train a RNN for multi class classification of faults from a number of selected features.
 
Can someone give me some advice if i'm doing things right? I'd really like to deploy the whole system with caffe2 but right now i can't see progresses in the training.
My net is defined as
```python
        self.model = model_helper.ModelHelper(name="rnn")
        
        self.input_blob, self.seq_lengths, self.hidden_init, self.cell_init, self.distance_vector, self.zeros = \
            self.model.net.AddExternalInputs(
                'input_blob',
                'seq_lengths',
                'hidden_init',
                'cell_init',
                'distance_vector',
                'zeros'
            )
        
        # LSTM
        self.hidden_output_all, self.hidden_output, self.cell_t_all, self.cell_state = LSTM(
            self.model, self.input_blob, self.seq_lengths, (self.hidden_init, self.cell_init),
            self.D, self.hidden_size, scope="LSTM")
        
        # Reshape output
        self.output_reshaped, _ = self.model.net.Reshape(self.hidden_output_all, \
            ['output_reshaped', '_'], shape=[-1, self.hidden_size])
        
        
        # ReLU Prediction
        self.pred = self.model.net.Relu(self.output_reshaped, "pred")
        

        # Squared L2 Distance
        self.dist = self.model.net.SquaredL2Distance([self.target_vector, self.pred], 'dist')
                
        # Avereged Loss 
        self.loss = self.model.net.AveragedLoss([self.dist], "loss")
        
        self.model.AddGradientOperators([self.loss])
        
        build_sgd(
            self.model,
            base_learning_rate=0.01,
            policy="step",
            stepsize=1,
            gamma=0.9
        )
```

After that i initialize the net with
```python
        self.prepare_state = core.Net("prepare_state")
        self.prepare_state.Copy(self.hidden_output, self.hidden_init)
        self.prepare_state.Copy(self.cell_state, self.cell_init)
        
        workspace.RunNetOnce(self.model.param_init_net)
        
        workspace.FeedBlob(self.hidden_output, np.zeros([1, self.batch_size, self.hidden_size], dtype=np.float32))
        workspace.FeedBlob(self.cell_state, np.zeros([1, self.batch_size, self.hidden_size], dtype=np.float32))
        workspace.FeedBlob(self.zeros, np.zeros([self.batch_size*self.seq_len, 1], dtype=np.float32))
        workspace.CreateNet(self.prepare_state)
    
        workspace.RunNetOnce(self.prepare_state)
```

and then for every batch
```python
        workspace.FeedBlob('input_blob', self.data)
        workspace.FeedBlob("target_vector", self.target_vector)
                
        if first: #given parameter
            workspace.CreateNet(self.model.net) 
            
        workspace.RunNet(self.model.net.Name())
```
But the losses i read are or fixed or diverging at every epoch. (as example on last run i get 297, 511, 770, ..) as losses.
Am i doing something wrong? Can someone help me?

**Comment by danigamba** [comment]
This is wrong in several way. By the way a little more documentation on basic network creation, train and export would be nice

**Comment by kevinlin311tw** [comment]
Did you solve the problem? I am also wondering how to train multi-class classification using caffe2.

**Comment by danigamba** [comment]
I do multi-class classification using a CNN and SquaredL2Distance as loss function. There are other issues explaining it, but I didn't get it working with RNN

**Comment by CarlosYeverino** [comment]
Hello @danigamba ,

how did you manage to use a SquaredL2Distance as loss function in CNN models?
